### PR TITLE
Prevent duplication of user prompts / chat history in ChatSession.

### PR DESCRIPTION
**Pull request by philippjbauer**
The way ChatSession.ChatAsync was using the provided methods from a IHistoryTransform interface implementation created unexpected duplication of the chat history messages. It also prevented loading previous history into the current session.

### CODE REVIEW

File: LLama/ChatSession.cs [hunk 1]
```diff
@@ -135,30 +135,54 @@ public virtual void LoadSession(string path)
...
+        /// Generates a response for a given user prompt and manages history state for the user.
+        /// This will always pass the whole history to the model. Don't pass a whole history
+        /// to this method as the user prompt will be appended to the history of the current session.
+        /// If more control is needed, use the other overload of this method that accepts a ChatHistory object.
         /// </summary>
         /// <param name="prompt"></param>
         /// <param name="inferenceParams"></param>
         /// <param name="cancellationToken"></param>
-        /// <returns></returns>
+        /// <returns>Returns generated tokens of the assistant message.</returns>
```
Comment (martindevans) [inline LLama/ChatSession.cs:146]:
```suggestion
        /// <returns>Returns generated text of the assistant message.</returns>
```

File: LLama/ChatSession.cs [hunk 2]
```diff
@@ -135,30 +135,54 @@ public virtual void LoadSession(string path)
...
         {
-            foreach(var inputTransform in InputTransformPipeline)
+            foreach (var inputTransform in InputTransformPipeline)
                 prompt = inputTransform.Transform(prompt);
-            
-            History.Messages.AddRange(HistoryTransform.TextToHistory(AuthorRole.User, prompt).Messages);
+
+            History.Messages.Add(new ChatHistory.Message(AuthorRole.User, prompt));
+
+            string internalPrompt = HistoryTransform.HistoryToText(History);
```
Comment (martindevans) [inline LLama/ChatSession.cs:154]: This looks like it prompts with the _entire history_ every time, that doesn't seem correct? Unless you're using a stateless executor then all of the history is still stored within the context KV cache.

**Comment by martindevans** [comment]
Note: refer to #261 for some initial discussion

File: LLama/ChatSession.cs (hunk 2 above)
Comment (philippjbauer) [reply LLama/ChatSession.cs]: That was not the intention. I updated the master branch to only use the full history the first time it runs. This enables the use/developer to prime the conversation with a system message and a given back and forth between the user and assistant before handing it off to the first chat message the user enters.

**Review by martindevans — APPROVED** [review]
LGTM!

**Comment by AsakusaRinne** [comment]
LGTM, thank you very much for this contribution!

**Comment by AsakusaRinne** [comment]
> Glad to help out! I hope to find some time to help advance the overall ChatSession class but I've got a pretty packed calendar in the next few weeks.

No problem, you're always welcome :) If you'd like to take part in the development of LLamaSharp, I'll invite you to the discord dev channel.

**Pull request merged**